frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,372 @@
+}
+
+sbp('chelonia/defineContract', {
+  name: 'gi.contracts/chatroom',
+  metadata: {
+    validate: objectOf({
+      createdDate: string, // action created date
+      username: string, // action creator
+      identityContractID: string // action creator identityContractID
+    }),
+    create () {
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
+      return {
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
+      }
+    }
+  },
+  state (contractID) {
+    return sbp('state/vuex/state')[contractID]
+  },
+  getters: {
+    currentChatRoomState (state) {
+      return state
+    },
+    chatRoomSettings (state, getters) {
+      return getters.currentChatRoomState.settings || {}
+    },
+    chatRoomAttributes (state, getters) {
+      return getters.currentChatRoomState.attributes || {}
+    },
+    chatRoomUsers (state, getters) {
+      return getters.currentChatRoomState.users || {}
+    },
+    chatRoomLatestMessages (state, getters) {
+      const messages = getters.currentChatRoomState.messages || []
+      return getLatestMessages({
+        count: getters.chatRoomSettings.messagesPerPage,
+        messages
+      })
+    }
+  },
+  actions: {
+    // This is the constructor of Chat contract
+    'gi.contracts/chatroom': {
+      validate: chatRoomType,
+      process ({ meta, data }, { state }) {
+        const initialState = merge({
+          settings: {
+            messagesPerPage: CHATROOM_MESSAGES_PER_PAGE,
+            maxNameLetters: CHATROOM_NAME_LIMITS_IN_CHARS,
+            maxDescriptionLetters: CHATROOM_DESCRIPTION_LIMITS_IN_CHARS
+          },
+          users: {},
+          messages: []
+        }, {
+          attributes: {
+            ...data,
+            creator: meta.username,
+            deletedDate: null,
+            archievedDate: null
+          }
+        })
+        for (const key in initialState) {
+          Vue.set(state, key, initialState[key])
+        }
+      },
+      sideEffect ({ contractID, meta }) {
+        const rootState = sbp('state/vuex/state')
+        if (meta.username === rootState.loggedIn.username) {
+          sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: contractID })
+        }
+      }
+    },
+    'gi.contracts/chatroom/join': {
+      validate: objectMaybeOf({
+        username: string,
+        referer: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        const { username, referer } = data
+
+        const notificationType = username === referer ? MESSAGE_NOTIFICATIONS.JOIN_MEMBER : MESSAGE_NOTIFICATIONS.ADD_MEMBER
+        const notificationData = createNotificationData(
+          notificationType,
+          notificationType === MESSAGE_NOTIFICATIONS.ADD_MEMBER ? { username } : {}
+        )
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        Vue.set(state.messages, [state.messages.length], newMessage)
```

Comment:
Arrays do not need `Vue.set`, that is only necessary for Objects. Vue.js should handle reactivity and updating UI with `state.message.push(newMessage)`

Reply:
Make sense.